### PR TITLE
feat: multiple_facet knowledge_panel

### PR DESCRIPTION
**Pull request by sumit-158**
### What
- Build multiple facet kp for last-edits, data-quality and hunger games

### Fixes bug(s)
- #19

### CODE REVIEW

File: tests/test_knowledge_panels.py [hunk 1]
```diff
@@ -64,44 +64,20 @@ def test_hunger_game_kp_category_with_country():
...
-                    },
-                }
-            ]
-        }
-    }
-
-
-def test_hunger_game_kp_label_with_value():
-    assert hunger_game_kp(hunger_game_filter="label", value="organic") == {
+def test_hunger_game_kp_category_with_tags():
```
Comment (alexgarel) [inline tests/test_knowledge_panels.py:67]: why did you remove so many tests ?

- test_hunger_game_kp_category_with_value
- test_hunger_game_kp_brand_with_value
- test_hunger_game_kp_label_with_value

File: app/knowledge_panels.py [hunk 2]
```diff
@@ -23,6 +25,11 @@ def hunger_game_kp(
     if value is not None:
         query["value_tag"] = value
         description = f"{value} {hunger_game_filter}"
+    if sec_facet is not None:
+        description += f" {sec_facet}"
+    if sec_value is not None:
+        query[sec_facet] = sec_value
+        description += f" {sec_value}"
```
Comment (alexgarel) [inline app/knowledge_panels.py:32]: This does not work and won't do anything.
The only case where we could really cross values is the case were one of both facet is country or brand.

So you could handle both those cases (remember: first unify your cases, then process).

And when you have two non "mixable" facets, I would propose to put a link for each facet/value instead.

I would say:
- receive facet name and values as a dict, with None for value if you only have a facet name
- if there is a "country" parameter, remove from dict and use it to filter on country
- if there is a "brand" parameter, remove from dict and use it to filter on brand
- then if you still have two value in dict, generates two panels
- if you have no more value, and got a brand parameter, ask for brand questions
- else generate only one panel

File: app/knowledge_panels.py [hunk 3]
```diff
@@ -118,13 +136,22 @@ def last_edits_kp(
     if value is not None:
         query[f"{facet_plural(facet=facet)}_tags_en"] = value
         description += f" {value}"
+        source_url = f"{url}/{facet}/{value}?sort_by=last_modified_t"
+    if sec_facet is not None:
+        description += f" {sec_facet}"
```
Comment (alexgarel) [inline app/knowledge_panels.py:141]: add it to description only  if sec_value is not None (because otherwise it does not mean much, as it has no impact on search)

File: tests/test_knowledge_panels.py [hunk 4]
```diff
@@ -64,44 +64,20 @@ def test_hunger_game_kp_category_with_country():
...
+        sec_value="coffees",
+        country="france",
+    ) == {
         "hunger-game": {
             "elements": [
                 {
                     "element_type": "text",
                     "text_element": {
-                        "html": "<p><a href='https://hunger.openfoodfacts.org/questions?type=label&value_tag=organic'>Answer robotoff questions about organic label</a></p>\n"
+                        "html": "<p><a href='https://hunger.openfoodfacts.org/questions?country=france&type=brand&value_tag=nestle&category=coffees'>Answer robotoff questions about nestle brand category coffees</a></p>\n"
```
Comment (alexgarel) [inline tests/test_knowledge_panels.py:80]: as said above, you will see that this does not filter on category coffees (hunger game does not support that yet !).

Also trying it, I found another bug: it seems you now have to prefix country with "en:" for the filter to work. You should change your function to handle that.

**Review by alexgarel — CHANGES_REQUESTED** [review]
Two panels are ok, great.
But there are changes to be made for one of them, and for hunger game, it have to change much !

File: tests/test_knowledge_panels.py (hunk 4 above)
Comment (sumit-158) [reply tests/test_knowledge_panels.py:80]: got it!

File: tests/test_knowledge_panels.py (hunk 1 above)
Comment (sumit-158) [reply tests/test_knowledge_panels.py]: Added!

File: app/knowledge_panels.py (hunk 2 above)
Comment (sumit-158) [reply app/knowledge_panels.py]: @alexgarel I use other approach!

File: app/knowledge_panels.py [hunk 5]
```diff
@@ -16,14 +18,25 @@ def hunger_game_kp(
...
+                description += f" {sec_facet} {sec_value}"
+        else:
+            query["type"] = f"{hunger_game_filter}"
+            description = f"{hunger_game_filter}"
+            if value is not None:
+                query["value_tag"] = value
+                description = f"{hunger_game_filter} {value}"
+            if sec_facet == "brand":
+                query[sec_facet] = sec_value
+                description += f" {sec_facet} {sec_value}"
```
Comment (alexgarel) [inline app/knowledge_panels.py:39]: This is quite close, but it's a bit messy.

Once again I will say, we have to unite and process.

Moreover I would expect that in the case we have two facets that are not mergable into one hunger game query, we return a double knowledge panel, with question for one facets and for the other.

If you are ok to try, I would say:
1. first unite all facets in a dict `facets = {hunger_game_filter: value, sec_facet: sec_value}`
2. eventually remove [None] keys and [None] values (it does not have sense in hungergame to filter without a value)
3. if there is a "brand" key, use it as a global filter and pop it out from the dict
4. for each key, value remaining in the dict build the question url, eventually adding brand filter, if there is one
5 build the knowledge panel

Of course you will have to change tests and try different test cases.

File: app/main.py [hunk 6]
```diff
@@ -31,16 +33,38 @@ def knowledge_panel(
         panels = []
         if facet_tag in HungerGameFilter.list():
             panels.append(
-                hunger_game_kp(hunger_game_filter=facet_tag, value=value_tag, country=country)
+                hunger_game_kp(
+                    hunger_game_filter=facet_tag,
+                    value=value_tag,
+                    sec_facet=sec_facet_tag,
+                    sec_value=sec_value_tag,
```
Comment (alexgarel) [inline app/main.py:40]: It would be better to stick to same parameters name for every function. (see polymorphism, that is having same interface, enables nice patterns)

So I would rename hunger_game_filter and value to facet and value.

Guess what ? Then you could have a loop here ;-)

File: app/main.py [hunk 7]
```diff
@@ -31,16 +33,38 @@ def knowledge_panel(
...
+                    country=country,
+                )
+            )
         except Exception:
-            logging.exception("error occued while appending last-edits-kp")
+            logging.exception("error occued while appending data-quality-kp")
         try:
             if facet_tag in Taxonomies.list():
                 panels.append(wikidata_kp(facet=facet_tag, value=value_tag))
```
Comment (alexgarel) [inline app/main.py:70]: maybe here we would just like to generate two panels, one for each facet. We may need to specify the "id" we want to avoid having two panels with the same name ?

**Review by alexgarel — CHANGES_REQUESTED** [review]
You have done part of the road, but hold on :-)

File: app/knowledge_panels.py (hunk 5 above)
Comment (alexgarel) [reply app/knowledge_panels.py:39]: oh sorry, you're right, keep them :-)
I let you handle the logic in the above schema.

File: app/knowledge_panels.py [hunk 8]
```diff
@@ -1,167 +1,184 @@
...
-    return {
-        "hunger-game": {
-            "elements": [
+class KnowledgePanels:
+    def __init__(self, facet, value, sec_facet, sec_value, country):
+        self.facet = facet
+        self.value = value
+        self.sec_facet = sec_facet
+        self.sec_value = sec_value
+        self.country = country
```
Comment (sumit-158) [inline app/knowledge_panels.py:14]: Is this coorect?

Comment (alexgarel) [reply app/knowledge_panels.py:14]: Oh I didn't think of it, I talked about polymorphism as a comparison. But that's a good idea indeed, I like it :-)

Comment (sumit-158) [reply app/knowledge_panels.py:14]: How to test it though?

File: app/knowledge_panels.py [hunk 9]
```diff
@@ -1,167 +1,191 @@
...
+            description += f"{self.facet}"
+        if self.value is not None:
+            query[f"{facet_plural(facet=self.facet)}_tags_en"] = self.value
+            description += f" {self.value}"
+            source_url = f"{url}/{self.facet}/{self.value}?sort_by=last_modified_t"
+        if self.sec_value and self.sec_facet is not None:
+            query[f"{facet_plural(facet=self.sec_facet)}_tags_en"] = self.sec_value
+            description += f" {self.sec_facet} {self.sec_value}"
+            source_url = f"{url}/{self.facet}/{self.value}/{self.sec_facet}/{self.sec_value}?sort_by=last_modified_t"  # noqa: E501
+        expected_html, t_description, t_title = last_edit(url=url, query=query)
```
Comment (alexgarel) [inline app/knowledge_panels.py:145]: minor:
expected_html is a strange name (it's for test that we use this expression).
It's more, edits_html or t_html.

File: app/knowledge_panels.py [hunk 10]
```diff
@@ -1,167 +1,191 @@
...
-            "source_url": f"{url}/{facet}/{value}?sort_by=last_modified_t",
-            "elements": [
-                {
-                    "element_type": "text",
-                    "text_element": expected_html,
-                },
-            ],
-        },
-    }
+    def wikidata_kp(self):
```
Comment (alexgarel) [inline app/knowledge_panels.py:161]: you have to make wikidata for both facet/value.

A good way to implement that is to have a private method _wikidata_kp(self, facet, value)
and to call it twice (once with main facet, once with second facet) in wikidata_kp, and compose results.

**Review by alexgarel — CHANGES_REQUESTED** [review]
You're code looks beautiful, I can see progresses :-)

But there are still changes to be made to this PR.

File: app/knowledge_panels.py [hunk 11]
```diff
@@ -1,167 +1,191 @@
...
+        facets.update(filtered)
+        urls = set()
+        html = []
+        if self.country is not None:
+            query["country"] = f"en:{self.country}"
+        if "country" in facets.keys():
+            facet_value = facets.get("country")
+            query["country"] = f"en:{facet_value}"
+            facets.pop("country")
+        if "brand" in facets.keys():
```
[the remark this answers is not in the document]
Comment (sumit-158) [reply app/knowledge_panels.py:38]: @alexgarel but the current code tackles that problem, if brand is a sec facet and we have a value and also if we only have brand

[the remark this answers is not in the document]
Comment (alexgarel) [reply app/knowledge_panels.py:40]: yes it's another problem, that you go into, you should not pop brand if it's the only facet, that's why I propose this solution, also you should not pop it if it does not have a value, your code pops it anyway.

File: app/knowledge_panels.py [hunk 12]
```diff
@@ -1,167 +1,191 @@
...
+            facets.pop("brand")
+        for k, v in facets.items():
+            query["type"] = k
+            if v is not None:
+                query["value_tag"] = v
+            urls.add(questions_url + f"?{urlencode(query)}")
+        if len(facets) != 2:
+            questions_url += f"?{urlencode(query)}"
+            urls.add(questions_url)
+        t_description = hungergame()
```
[the remark this answers is not in the document]
Comment (alexgarel) [reply app/knowledge_panels.py:55]: If you feel it's too much work for now to do it, put a TODO comment about it.

File: tests/test_knowledge_panels.py [hunk 13]
```diff
@@ -19,114 +18,185 @@ def auto_activate_lang():
...
+        country="belgium",
+    ).hunger_game_kp() == {
+        "hunger-game": {
+            "title": "hunger-games",
+            "elements": [
+                {"id": 0, "element_type": "text", "text_element": {"html": html0}},
+                {"id": 1, "element_type": "text", "text_element": {"html": html1}},
+            ],
+        }
+    }
```
Comment (sumit-158) [inline tests/test_knowledge_panels.py:175]: @alexgarel `test_hunger_game_kp_with_all_tag_2 and 3` is failing due to change in  value `html0` `html1` while adding urls, in `knowledge_panels.py`

File: app/knowledge_panels.py [hunk 14]
```diff
@@ -1,167 +1,244 @@
...
+                urls.add((f"{questions_url}?{urlencode(query)}", description))
+        elif len(facets) >= 1:
+            for k, v in facets.items():
+                query["type"] = k
+                description = k
+                if v is not None:
+                    query["value_tag"] = v
+                    description += f" {v}"
+                urls.add((f"{questions_url}?{urlencode(query)}", description))
+        if query:
```
Comment (sumit-158) [inline app/knowledge_panels.py:72]: here!

Comment (sumit-158) [reply app/knowledge_panels.py:72]: It adds url unorderly.

File: tests/test_knowledge_panels.py (hunk 13 above)
Comment (sumit-158) [reply tests/test_knowledge_panels.py:175]: Any suggestion?

Comment (sumit-158) [reply tests/test_knowledge_panels.py:175]: It's like sometimes `html0` became `html1` and vice-versa.

Comment (sumit-158) [reply tests/test_knowledge_panels.py:175]: here,
image

File: app/knowledge_panels.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (alexgarel) [reply app/knowledge_panels.py]: you did not do what I asked, you do the contrary !

File: app/knowledge_panels.py [hunk 15]
```diff
@@ -1,167 +1,191 @@
...
+        if "country" in facets.keys():
+            facet_value = facets.get("country")
+            query["country"] = f"en:{facet_value}"
+            facets.pop("country")
+        if "brand" in facets.keys():
+            facet_value = facets.get("brand")
+            if facet_value is not None:
+                query["brand"] = facet_value
+            else:
+                query["type"] = "brand"
```
[the remark this answers is not in the document]
Comment (alexgarel) [reply app/knowledge_panels.py:47]: You did not take into account my comment !

Here https://github.com/openfoodfacts/facets-knowledge-panels/pull/27#pullrequestreview-1103654180 I told you to straighten the condition above (you did rewrite it, but on the contrary of what was to be done…).

If you write the above if as proposed, in this piece of code, you will just have to add brand_value as query["brand"].
All the rest will be solved by the underneath loop, without the need to all the specific code you added underneath…

File: app/knowledge_panels.py [hunk 16]
```diff
@@ -1,167 +1,244 @@
...
+                    description += f" {v}"
+                urls.add((f"{questions_url}?{urlencode(query)}", description))
+        elif len(facets) >= 1:
+            for k, v in facets.items():
+                query["type"] = k
+                description = k
+                if v is not None:
+                    query["value_tag"] = v
+                    description += f" {v}"
+                urls.add((f"{questions_url}?{urlencode(query)}", description))
```
Comment (alexgarel) [inline app/knowledge_panels.py:71]: This should simply be rewritten:

```suggestion
        brand_description = ""
        if facets.get("brand") and len(facets) > 1:
            # we have a brand and other facets, let's use brand as a filter
            brand_value = facets.pop("brand")
            query["brand"] = brand_value
            brand_description = f" for brand {brand_value}"
        # generate an url for each remaining facets
        for k, v in facets.items():
            query["type"] = k
            facet_description = k
            if v is not None:
                query["value_tag"] = v
                description += f" {v}"
            urls.add((
                f"{questions_url}?{urlencode(query)}", 
                f"questions about {facet_description}{brand_description}{description}"
            ))
```

See how simple it is :-)

Comment (alexgarel) [reply app/knowledge_panels.py:71]: I changed it a bit, because we have to copy query before adding type and value.
If not, if you have first facet with a value and second without a value, the first facet value would have been used for the second facet.

File: tests/test_knowledge_panels.py (hunk 13 above)
Comment (alexgarel) [reply tests/test_knowledge_panels.py:175]: hum, indeed this is because we puts facets in a dict (which does not have a predictable order).

I would modify the function, instead of `for k, v in facets.items()` to `for k, v in sorted(facets.items())` for it's predictable.

File: app/knowledge_panels.py (hunk 14 above)
Comment (alexgarel) [reply app/knowledge_panels.py:72]: see https://github.com/openfoodfacts/facets-knowledge-panels/pull/27#discussion_r971133767

File: app/knowledge_panels.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (sumit-158) [reply app/knowledge_panels.py:47]: My bad! I thought I had to have just a single KP for brand and another facet, that is why I was doing that, anyway thank you for the guidance, I pushed your suggestions with little change in des. I guess now it's ready to merge :)

File: app/knowledge_panels.py (hunk 16 above)
Comment (sumit-158) [reply app/knowledge_panels.py:71]: Thank you!

File: app/knowledge_panels.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (alexgarel) [reply app/knowledge_panels.py]: Oh, I understand better ! Sorry if I did not express myself well enough.

**Review by alexgarel — APPROVED** [review]
Let's merge it :-)
